app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt

```diff
@@ -0,0 +1,126 @@
+            }.flatMapLatest { apiMovies ->
+                movieDao.getAll()
+                    .map { favoriteMovies ->
+                        apiMovies.map { apiMovie ->
+                            apiMovie.toMovie(isFavorite = favoriteMovies.any { it.id == apiMovie.id })
+                        }
+                    }
+            }.shareIn(
+                scope = CoroutineScope(bgDispatcher),
+                started = SharingStarted.WhileSubscribed(1000L),
+                replay = 1
+            )
+        }
+
+    private val favorites = movieDao.getAll().map {
+        it.map { dbFavoriteMovie ->
+            Movie(
+                id = dbFavoriteMovie.id,
+                imageUrl = dbFavoriteMovie.posterUrl,
+                title = "",
+                overview = "",
+                isFavorite = true
+            )
+        }
+    }.shareIn(
+        scope = CoroutineScope(bgDispatcher),
+        started = SharingStarted.WhileSubscribed(1000L),
+        replay = 1
+    )
+
+
+    override fun popularMovies(movieCategory: MovieCategory): Flow<List<Movie>> =
+        moviesByCategory[movieCategory]!!
+
+    override fun nowPlayingMovies(movieCategory: MovieCategory): Flow<List<Movie>> =
+        moviesByCategory[movieCategory]!!
+
+    override fun upcomingMovies(movieCategory: MovieCategory): Flow<List<Movie>> =
+        moviesByCategory[movieCategory]!!
+
+    override fun movieDetails(movieId: Int): Flow<MovieDetails> = flow {
+        emit(movieService.fetchMovieDetails(movieId) to movieService.fetchMovieCredits(movieId))
+    }.flatMapLatest { (apiMovieDetails, apiMovieCredits) ->
+        movieDao.getAll()
+            .map { favoriteMovies ->
+                apiMovieDetails.toMovieDetails(
+                    isFavorite = favoriteMovies.any { it.id == apiMovieDetails.id },
+                    crew = apiMovieCredits.crew.map { it.toCrewman() },
+                    cast = apiMovieCredits.cast.map { it.toActor() }
+                )
+            }
+    }.flowOn(bgDispatcher)
+
+    override fun favoriteMovies(): Flow<List<Movie>> = favorites
+
+    override suspend fun addMovieToFavorites(movieId: Int) {
+        val movie = findMovie(movieId) ?: return
+        movie.imageUrl?.let {
+            DbFavoriteMovie(movieId, it).let { dbFavoriteMovie ->
+                movieDao.insert(dbFavoriteMovie)
+            }
+        }
+
+    }
+
+    private suspend fun findMovie(movieId: Int): Movie? {
+        var movie: Movie? = null
+        moviesByCategory.values.forEach { value ->
+            val movies = value.first()
+            movies.forEach {
+                if (it.id == movieId) {
+                    movie = it
+                }
+            }
+
+        }
+        return movie
+    }
+
+    override suspend fun removeMovieFromFavorites(movieId: Int) = movieDao.delete(movieId)
+
+    override suspend fun toggleFavorite(movieId: Int) {
+        runBlocking(bgDispatcher) {
```

Comment:
why runBlocking when this function is already suspended - change dispatcher on invocation in ViewModel